### PR TITLE
Don't allow exceptions to be swallowed when parsing optional fields

### PR DESCRIPTION
**Pull request by feihong**
This PR attempts to fix https://github.com/ahrefs/bs-atdgen-codec-runtime/issues/9

### CODE REVIEW

File: esy.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (feihong) [inline esy.json]: I added esy to this project, is it OK?

File: src/atdgen_codec_runtime.ml [hunk 2]
```diff
@@ -173,7 +173,36 @@ struct
...
+          DecodeError msg -> raise @@ DecodeError (msg ^ "\n\tat field '" ^ key ^ "'")
+        end
+      | None ->
+        raise @@ MissingFieldError ({j|Expected field '$(key)'|j})
+    end
+    else
+      raise @@ DecodeError ("Expected object, got " ^ Js.Json.stringify json)
+
+  (* We don't use Json_decode.optional because it swallows exceptions *)
+  let sane_optional decode json = Some (decode json)
```
Comment (ygrek) [inline src/atdgen_codec_runtime.ml]: `try .. catch -> None` should be in sane_optional (also maybe better name? and link to bs-json issue)

File: src/atdgen_codec_runtime.ml [hunk 3]
```diff
@@ -173,7 +173,36 @@ struct
     else
       Some (decode json)
 
-  let fieldOptional s f = optional (field s f)
+  exception MissingFieldError of string
```
Comment (ygrek) [inline src/atdgen_codec_runtime.ml]: do we really need new exception? can make `field` return `option`?

Comment (feihong) [reply src/atdgen_codec_runtime.ml]: Changed, see https://github.com/ahrefs/bs-atdgen-codec-runtime/pull/11/commits/1fb4750c124f7e459776bc6214de2bf952ecf55d#diff-af319977d60814aacb356044a4666058R176

File: src/atdgen_codec_runtime.ml (hunk 2 above)
Comment (feihong) [reply src/atdgen_codec_runtime.ml]: Renamed to custom_optional and added link to bs-json issue: https://github.com/ahrefs/bs-atdgen-codec-runtime/pull/11/commits/1fb4750c124f7e459776bc6214de2bf952ecf55d#diff-af319977d60814aacb356044a4666058R199

File: src/atdgen_codec_runtime.ml [hunk 4]
```diff
@@ -173,7 +173,35 @@ struct
     else
       Some (decode json)
 
-  let fieldOptional s f = optional (field s f)
+  (* Define our own version of field which returns None if key is not found *)
+  let custom_field key decode json =
```
Comment (ygrek) [inline src/atdgen_codec_runtime.ml]:
```suggestion
  let fieldOptional key decode json =
```

and custom_optional is not needed then

Comment (feihong) [reply src/atdgen_codec_runtime.ml]: Changed in https://github.com/ahrefs/bs-atdgen-codec-runtime/pull/11/commits/0549888c3e2635b4876a9c4a5239ae0670552712

**Comment by Khady** [comment]
thanks

**Pull request merged**